development/webpack/webpack.config.ts

```diff
@@ -0,0 +1,397 @@
+      buildDependencies: {
+        defaultConfig: [__filename],
+        // Invalidates the build cache when the listed files change.
+        // `__filename` makes all `require`d dependencies of *this* file
+        // `buildDependencies`
+        config: [
+          __filename,
+          join(context, '../.metamaskrc'),
+          join(context, '../builds.yml'),
+          browsersListPath,
+        ],
+      },
+    } as const satisfies FileCacheOptions)
+  : ({ type: 'memory' } as const satisfies MemoryCacheOptions);
+// #endregion cache
+
+// #region plugins
+const commitHash = isDevelopment ? getLatestCommit().hash() : null;
+const plugins: WebpackPluginInstance[] = [
+  new SelfInjectPlugin({ test: /^scripts\/inpage\.js$/u }),
+  // HtmlBundlerPlugin treats HTML files as entry points
+  new HtmlBundlerPlugin({
+    preprocessorOptions: { useWith: false },
+    minify: args.minify,
+    integrity: 'auto',
+    test: /\.html$/u, // default is eta/html, we only want html
+    data: {
+      isMMI: args.type === 'mmi',
+      isTest: args.test,
+      shouldIncludeSnow: args.snow,
+    },
+  }),
+  new ManifestPlugin({
+    web_accessible_resources: webAccessibleResources,
+    manifest_version: MANIFEST_VERSION,
+    description: commitHash
+      ? `${args.env} build from git id: ${commitHash.substring(0, 8)}`
+      : null,
+    version: variables.get('METAMASK_VERSION') as string,
+    browsers: args.browser,
+    transform: args.lockdown
+      ? undefined
+      : (browserManifest) => {
+          // remove lockdown scripts from content_scripts
+          const mainScripts = browserManifest.content_scripts?.[0];
+          if (mainScripts) {
+            const keep = ['scripts/contentscript.js', 'scripts/inpage.js'];
+            mainScripts.js = mainScripts.js?.filter((js) => keep.includes(js));
+          }
+          return browserManifest;
+        },
+    zip: args.zip,
+    ...(args.zip
+      ? {
+          zipOptions: {
+            outFilePath: `../../../builds/metamask-[browser]-${variables.get(
+              'METAMASK_VERSION',
+            )}.zip`, // relative to output.path
+            mtime: getLatestCommit().timestamp(),
+            excludeExtensions: ['.map'],
+            // `level: 9` is the highest; it may increase build time by ~5% over level 1
+            level: 9,
+          },
+        }
+      : {}),
+  }),
+  // use ProvidePlugin to polyfill *global* node variables
+  new ProvidePlugin({
+    // Make a global `Buffer` variable that points to the `buffer` package.
+    Buffer: ['buffer', 'Buffer'],
+    // Make a global `process` variable that points to the `process` package.
+    process: 'process/browser',
+  }),
+  new CopyPlugin({
+    patterns: [
+      { from: join(context, '_locales'), to: '_locales' }, // translations
+      // misc images
+      // TODO: fix overlap between this folder and automatically bundled assets
+      { from: join(context, 'images'), to: 'images' },
```

Comment:
Not all of our assets are automatically discoverable from static analysis alone, so we still need to copy things.

_Some_ are discoverable though, and those get automatically bundled (to a different directory), which does result in some duplication.